### ISSUE TITLE
Allow --passWithNoTests

### ISSUE THREAD

**Issue by SleeplessByte** [tstyche/tstyche#268]
For CI usage, there are cases where no tests are fine. There is currently no easy way to make the CLI pass when there are no tests. It will always yield

> No test files were selected using current configuration.

The relevant option for `jest` is [`--passWithNoTests`](https://jestjs.io/docs/cli#--passwithnotests). It would be great if tstyche could also support this.

**Comment by mrazauskas** [comment]
Thanks for feedback.

Looks like I forgot to add a note to documentation. Simply add the following to TSTyche config file:

```jsonc
{
  "$schema": "./models/config-schema.json",
  // TODO specify after adding the test files
  "testFileMatch": []
}
```

When `testFileMatch` patterns are specified (explicitly or picked from defaults), TSTyche does its best to find the files. That’s why the error says: "No test files were selected..."

An empty list makes it clear that there are no files to select. The only limitation is that `testFileMatch` cannot be passed as a CLI flag. Is there any reason you would like to use a flag instead?

---

In fact this option was implemented, but I removed it. This is because `test()` is required in Jest, but not in TSTyche. The `expect` assertions throw in Jest and that is why `test()` is required to catch those. Since files are only statically analysed in TSTyche, assertions do not throw. Hence neither `test()`, nor `expect` are required.

TSTyche would simply type check any files selected by `testFileMatch` patterns. Those can be empty files as well.

**Comment by SleeplessByte** [comment]
I volunteer as maintainer for exercism.org (completely free platform to teach programmers new programming languages). Among JavaScript and TypeScript, the platform supports over 60 other programming languages, but I mostly work on the first two.

Recently I have started concepts for exercises with type tests. First I wanted to use `tsd` as it has a jest runner (using tsd-lite) but as you announced... it is deprecated. The _reason_ I wanted to use that is because I wrote a reporter for jest that can output the jest output (json) to a format Exercism understands. The current iteration of this project is https://github.com/exercism/typescript-test-runner.

Students are able to complete solutions online (in an editor) that runs this test runner in a readonly and offline docker instance, or they can download the solution + tests to their machine and develop locally. A student may use any platform (Windows, Linux, MacOSX) and any tool. This means that we try to provide cross-platform tooling that's easy to run.

```shell
corepack yarn test
```

This is what a student runs (see it defined [in an exercise](https://github.com/exercism/typescript/blob/main/exercises/concept/lasagna/package.json#L30-L37)), and this will delegate to `tsc`, `tstyche` and `jest`. The first use case is here. I would prefer `tsc && tstyche --okay-if-there-are-no-tests && jest`. Right now I can't do that and have to delegate to a separate script that [determines if it should run tstyche](https://github.com/exercism/typescript/blob/main/exercises/concept/lasagna/.meta/test-runner.mjs). I do have it working so I technically no longer need the requested functionality.

Given the configuration option, why don't I just use that? Well that's where the second use case comes in. The test-runner should run both on new solutions (which would have this configuration change) as well as older solutions (which would not yet have this configuration change). I could late-inject the configuration, but that's as hacky as using a separate script to "maybe run tstyche".

As you can see, whilst not impossible, there is quite a bit of friction to run tstyche and accept 0 checks ran.

Let me know if you want more information.

**Comment by mrazauskas** [comment]
Thanks. This is very interesting use case.

How do you proceed with the output? I mean, JSON report is also missing in TSTyche.

What about running TSTyche [programatically](https://tstyche.org/reference/programmatic-usage). I am building an extension for VS Code and it sounds like you could simply use the same APIs. This means it is your responsibility to provide valid config and you can get back the results as plain JavaScript objects from an event emitter.

In other words, all you need is a custom reporter and a simple wrapper similar to this class: https://github.com/tstyche/tstyche/blob/main/source/api/TSTyche.ts

A reporter that only cares about the final result looks like this: https://github.com/tstyche/tstyche/blob/main/source/handlers/SummaryReporter.ts

As you see, it must wait for the `run:end` event that has `result` in its payload. The `Result` shape can be seen here: https://github.com/tstyche/tstyche/blob/main/source/result/Result.ts

---

This is just a quick demo of what is possible. Just wanted to check how you feel about that. If it sounds like a good solution, I can provide more details. I wanted to add this to the [Programmatic Usage](https://tstyche.org/reference/programmatic-usage) page at some point.

There is a note that programatic APIs are not ready for public use yet. Ignore that (; What I mentioned here can be used.

**Comment by SleeplessByte** [comment]
> In other words, all you need is a custom reporter and a simple wrapper similar to this class:

Fun fact, this is [_exactly what we do for jest_](https://github.com/exercism/typescript-test-runner/blob/main/src/reporter.ts#L11). So doing this would be preferable to me. You are perhaps interested in what we output so I will list that down below.

> A reporter that only cares about the final result looks like this:

That's fine. I would output a second file there so can even extend the original one.

> There is a note that programatic APIs are not ready for public use yet. Ignore that (; What I mentioned here can be used.

Yeah I stopped reading when I saw that and didn't investigate further.

----------

> How do you proceed with the output? I mean, JSON report is also missing in TSTyche.

It's cumbersome.

Let's assume a test file with at least the following content:

```typescript
// import { ... }

describe('EXPECTED_MINUTES_IN_OVEN', () => {
  test('constant is defined as a number or a constant number', () => {
    expect(EXPECTED_MINUTES_IN_OVEN).type.toBeAssignableTo<number>()
  })
})

// [...]
```

- Read each line of `stdout` after subtracting `stderr`.
- Then read each `+` or `x` so I know which things have passed or failed
- Then read the code with [this static analysis lib I wrote](https://github.com/exercism/javascript-lib-static-analysis/blob/main/src/extracts/extract_tests.ts) to extract the `describe` and `test` calls
- Now I can build the following output:

```json
{
  "name": "EXPECTED_MINUTES_IN_OVEN > constant is defined as a number or a constant number",
  "status": "pass",
  "test_code": "expect(EXPECTED_MINUTES_IN_OVEN).type.toBeAssignableTo<number>()",
  "message": "",
  "expected": null,
  "output": null,
  "task_id": 1
}
```

This output is then finally ingested by the platform and shown to a student:

![image](https://github.com/user-attachments/assets/2f36c4ed-1244-4c67-849b-7d2cf0cf8cda)

The fields `expected` and `output` are extra and tend to only be filled on failure, and the optimal one would be something like:

```json
{
  "name": "EXPECTED_MINUTES_IN_OVEN > constant is defined as a number or a constant number",
  "status": "fail",
  "test_code": "expect(EXPECTED_MINUTES_IN_OVEN).type.toBeAssignableTo<number>()",
  "message": "Error: \u001b[0mType 'unknown' is not assignable to type 'number'.\n\n\u001b[90m   9\u001b[0m\u001b[90m | \u001b[0mdescribe('EXPECTED_MINUTES_IN_OVEN', () => {\n\u001b[90m  10\u001b[0m\u001b[90m | \u001b[0m  test('constant is defined as a number or a constant number', () => {\n\u001b[31m  11\u001b[0m\u001b[90m | \u001b[0m    expect(EXPECTED_MINUTES_IN_OVEN).type.toBeAssignableTo<number>()\n    \u001b[90m | \u001b[0m                                                           \u001b[31m~~~~~~\u001b[0m\n\u001b[90m  12\u001b[0m\u001b[90m | \u001b[0m  })\n\u001b[90m  13\u001b[0m\u001b[90m | \u001b[0m})\n\u001b[90m  14\u001b[0m\u001b[90m | \u001b[0m\n\n      \u001b[90m at \u001b[0m\u001b[36m./__typetests__/lasagna.tst.ts\u001b[0m\u001b[90m:11:60\u001b[0m\u001b[90m \u276d EXPECTED_MINUTES_IN_OVEN \u276d constant is defined as a number or a constant number\u001b[0m\n",
  "expected": "type.toBeAssignableTo<number>",
  "output": "'unknown'",
  "task_id": 1
}
```

but I don't need them for this to work.

Here is an example from a jest test failing but I think you understand how the above would be translated in the output (without setting `expected` and `output`, it just shows `message` verbatim):

![image](https://github.com/user-attachments/assets/7adde900-852f-4c0b-811e-3fdf90625eae)

**Comment by mrazauskas** [comment]
Thanks. Let me think. I will try to put some code draft for you.

---

By the way, in TSTyche there isn’t something like expected/received. Similarly to how TypeScript works, output is a diagnostic message. While building TSTyche, this was the eureka! moment. It must simply wrap around the concepts from TypeScript.

Before that I was working on a replacement for `jest-runner-tsd` (i.e. trying to satisfy Jest's requirements). There was a lot of friction, admittedly. Switching to a wrapper of TypeScript made it all go smooth. Many Jest's ideas were adopted (obviously), but TSTyche had to become independent test runner.

**Comment by mrazauskas** [comment]
Quick question: are the test file names known? I mean, is your code able to provide the test file names or TSTyche should do the magic to select them?

**Comment by SleeplessByte** [comment]
> By the way, in TSTyche there isn’t something like expected/received. Similarly to how TypeScript works, output is a diagnostic message. While building TSTyche, this was the eureka! moment. It must simply wrap around the concepts from TypeScript.

Yep, I understood it to be statical analysis which I am used 😄 

> Quick question: are the test file names known? I mean, is your code able to provide the test file names or TSTyche should do the magic to select them?

This can be interpreted various ways:

1. Are you asking me if I can call the cli with the correct input? The answer to this is technically yes relying on glob so effectively no.
2. Are you asking me if I can call the programmatic entrypoint with the correct input? The answer to this is yes. It is stored in a config file [in a `.meta` directory](https://github.com/exercism/typescript/blob/main/exercises/concept/lasagna/.meta/config.json#L8) which will be read out anyway before calling the API.

For flexibility, I think the "if provided only do x, if not try to figure it out" is most flexible.

**Comment by mrazauskas** [comment]
Hm.. I have a feeling that currently programatic usage is really flexible, but for a simple use case (like yours) there is too much boilerplate code. It must be much more simple.

Working on that! Give me a couple of weeks, please (;

**Comment by mrazauskas** [comment]
Hey @SleeplessByte!

I have just published `tstyche@3.0.0-beta.4` with two new features: plugins and reporters.

Could you give it a try? I would like to collect some feedback before publishing a stable version.

---

Plugins allow to change behavior of TSTyche. For example, you can disable test file look up and provide own list of test files:

```js
// ./tstyche-plugin.js

/**
 * @type {import("tstyche/tstyche").Hooks}
 */
export default {
  config: (resolvedConfig) => {
    return { ...resolvedConfig, testFileMatch: [] /* disables look up */ };
  },
  select: async () => {
    return await getTestFilesSomehow();
  },
};
```

This is equivalent to `--passWithNoTests`. Plus you have a mechanism to provide test file list. It can be an empty array and will work: TSTyche will exit without any errors.

To use the plugin, run `tstyche --plugins './tstyche-plugin.js'`. 

Documentation: https://tstyche.org/guide/plugins

---

Reporters are simply event listeners. Full list of event can be found here: https://github.com/tstyche/tstyche/blob/main/source/events/types.ts

Not sure what would be the best in your case. In a way you could look at `run:end`, but its `Result` is huge. Perhaps it makes sense to listen for `task:` and `test:`?

```js
// ./tstyche-reporter.js

export default class CustomReporter {
  /**
   * @param {import("tstyche/tstyche").Event} event
   */
  on([event, payload]) {
    if (event === "run:start") {
      console.info("Hello from custom reporter!");

      for (const task of payload.result.tasks) {
        console.info(task.filePath);
      }
    }

    switch (event) {
      case "task:error":
        // there was an error, a test file (a or task) did not ran
        // no further events will be emitted
        break;

      case "task:end":
        // all tests in a file got evaluated
        break;

      case "test:error":
        // there was an error, test did not ran
        // no further events will be emitted
        break;

      case "test:fail":
        // test did not pass
        break;

      case "test:pass":
        // test passed
        break;

      case "test:skip":
        // test was skipped, but its inner structure was checked
        // any type errors are ignored
        break;

      case "test:todo":
        // test was marked todo and was ignored
        break;

      default:
        break;
    }
  }
}
```

I will work on events documentation in the upcoming days. Perhaps you could try to put together a PR? I could help to polish the reporter.

To use the reporter, run `tstyche --reporters './tstyche-reporter.js'`. 

Documentation: https://tstyche.org/guide/reporters

**Comment by mrazauskas** [comment]
By the way, a plugin also can load custom reporters:

```js
/**
 * @type {import("tstyche/tstyche").Hooks}
 */
export default {
  config: (resolvedConfig) => {
    return {
      ...resolvedConfig,
      reporters: [new URL("./custom-reporter.js", import.meta.url).toString()],
      testFileMatch: [] /* disables look up */,
    };
  },
};
```

**Comment by mrazauskas** [comment]
I have just published `tstyche@3.0.0-beta.6` with few minor changes (see updated the above examples):

- plugins are required to have `name` property;
- the plugin type is now called `Plugin`.

This must be the last beta before the rc release.

**Comment by SleeplessByte** [comment]
Due to illness I have not gotten around finishing my implementation with this, but I have been reading along and I think it's already "good enough" to work for what I am trying to accomplish.

**Comment by mrazauskas** [comment]
Sound good.

By the way, I forgot to mention that `select()` now also has context assigned to `this`. There is only one property for now: `this.resolvedConfig`. For instance, you could use `this.resolvedConfig.rootDir` if that is import.

I have to add a sentence about this detail to documentation (it is included in one example already) and the RC release will be ready.

**Comment by mrazauskas** [comment]
I have just shipped [tstyche@3.0.0](https://github.com/tstyche/tstyche/releases/tag/v3.0.0) with these newly added APIs. Hence I would like to close this issue as solved.

Thanks for opening this issue. Let’s continue talking about the implementation here. More than happy to answer any questions.
